src/signalhandler.cpp

```diff
@@ -0,0 +1,146 @@
+/******************************************************************************\
+ * Copyright (c) 2020
+ *
+ * Author(s):
+ *  Peter L Jones
+ *
+ ******************************************************************************
+ *
+ * This program is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation; either version 2 of the License, or (at your option) any later
+ * version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc.,
+ * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+ *
+\******************************************************************************/
+
+#include "signalhandler.h"
+
+class CSignalHandlerSingleton : public CSignalHandler {
+public:
+    inline CSignalHandlerSingleton() : CSignalHandler() {}
+};
+Q_GLOBAL_STATIC ( CSignalHandlerSingleton, singleton )
+
+CSignalHandler::CSignalHandler() : pSignalBase ( CSignalBase::withSignalHandler ( this ) ) {}
+
+CSignalHandler::~CSignalHandler() = default;
+
+CSignalHandler* CSignalHandler::getSingletonP() { return singleton; }
+
+bool CSignalHandler::emitSignal ( int sigNum )
+{
+    return QMetaObject::invokeMethod( singleton, "ShutdownSignal", Qt::QueuedConnection, Q_ARG( int, sigNum ) );
+}
+
+#ifndef _WIN32
+void CSignalHandler::OnSocketNotify( int socket )
+{
+    int sigNum;
+    if ( ::read ( socket, &sigNum, sizeof ( int ) ) == sizeof ( int ) )
+    {
+        emitSignal ( sigNum );
+    }
+}
+#endif
+
+// ----------------------------------------------------------
+
+CSignalBase::CSignalBase ( CSignalHandler* pSignalHandler ) :
+    pSignalHandler ( pSignalHandler )
+{
+}
+
+CSignalBase::~CSignalBase() = default;
+
+CSignalBase* CSignalBase::withSignalHandler ( CSignalHandler* pSignalHandler )
+{
+#ifdef _WIN32
+    return new CSignalWin ( pSignalHandler );
+#else
+    return new CSignalUnix ( pSignalHandler );
+#endif
+}
+
+#ifdef _WIN32
+CSignalWin::CSignalWin ( CSignalHandler* nPSignalHandler ) :
+    CSignalBase ( nPSignalHandler ),
+    lock ( QReadWriteLock::Recursive )
+{
+    SetConsoleCtrlHandler ( signalHandler, true );
+}
+
+CSignalWin::~CSignalWin() {
+    SetConsoleCtrlHandler ( signalHandler, false );
+}
+
+QReadWriteLock* CSignalWin::getLock() const
+{
+    return &lock;
+}
+
+BOOL WINAPI CSignalWin::signalHandler ( _In_ DWORD sigNum )
+{
+    auto self = getSelf<CSignalWin>();
+    QReadLocker lock ( self->getLock() );
+    return self->pSignalHandler->emitSignal ( static_cast<int>( sigNum ) );
+}
+
+#else
+int CSignalUnix::socketPair[2];
```

Comment:
Isn't this already declared in the header file?

Reply:
No - it's defined there.  This is the actual socket pair instance.  Two `int`s in an array, not "an array of `int`s".